rohmu/notifier/http.py

```diff
@@ -0,0 +1,132 @@
+"""Copyright (c) 2022 Aiven, Helsinki, Finland. https://aiven.io/"""
+from .interface import Notifier
+from dataclasses import dataclass
+from datetime import datetime
+from enum import Enum, unique
+from logging import getLogger
+from queue import SimpleQueue
+from requests.exceptions import RequestException
+from typing import Any, Dict, Optional
+
+import requests
+import threading
+
+LOG = getLogger(__name__)
+
+
+@unique
+class Operation(Enum):
+    UPLOAD = "UPLOAD"
+    DELETE = "DELETE"
+    DELETE_TREE = "DELETE_TREE"
+
+
+def _get_requests_session() -> requests.Session:
+    retry = requests.adapters.Retry(
+        total=3,
+        backoff_factor=0.5,
+        status_forcelist=[429, 500, 502, 503, 504],
+        method_whitelist={"POST"},
+    )
+    adapter = requests.adapters.HTTPAdapter(max_retries=retry)
+    request_session = requests.Session()
+    request_session.mount("http://", adapter)
+    return request_session
+
+
+class HTTPNotifier(Notifier):
+    def __init__(self, url: str) -> None:
+        self._session = _get_requests_session()
+        self._url = url
+
+    def object_created(self, key: str, size: Optional[int]) -> None:
+        self._session.post(
+            self._url,
+            {
+                "op": Operation.UPLOAD.value,
+                "key": key,
+                "size": size,
+                "last_modified": datetime.now(),
+            },
+        )
+
+    def object_deleted(self, key: str) -> None:
+        try:
+            self._session.post(self._url, {"op": Operation.DELETE.value, "key": key})
+        except RequestException as e:
+            LOG.warning(
+                "[HTTPNotifier] POST request to %r failed with %r, dropping event",
+                self._url,
+                str(e),
+            )
+
+    def tree_deleted(self, key: str) -> None:
+        try:
+            self._session.post(self._url, {"op": Operation.DELETE_TREE.value, "key": key})
+        except RequestException as e:
+            LOG.warning(
+                "[HTTPNotifier] POST request to %r failed with %r, dropping event",
+                self._url,
+                str(e),
+            )
+
+
+@dataclass
+class _HTTPNotifyJob:
```

Comment:
This should be `frozen=True` and the data field should be `Mapping` instead of `Dict`, especially since it crosses a threading boundary.

Actually, we shouldn't probably even have a generic `data`, making a dataclass but putting almost everything in a dict in a single field of the dataclass isn't great: serialization to HTTP-friendly but untyped format should be done as necessary when leaving the system, not before.

Reply:
> This should be frozen=True and the data field should be Mapping instead of Dict, especially since it crosses a threading boundary.

good point, done.

> Actually, we shouldn't probably even have a generic data, making a dataclass but putting almost everything in a dict in a single field of the dataclass isn't great: serialization to HTTP-friendly but untyped format should be done as necessary when leaving the system, not before.

I don't understand what you're trying to say. Are you proposing implementing a separate dataclass for each message type with (de)serialization? I would prefer to not do that since that would only increase code size without any benefits.

Reply:
Yes, I'm suggesting creating separate dataclass type for different message types :) Serialization can stay very simple and in a parent class though. It's mainly to avoid exchanging bundles of untyped dicts between parts of the code.

We've seen too many bugs caused by use of dictionaries that could have been caught by static type checks: missing key, wrong type of value, surprising None... Not having this type of bugs anymore would be the main benefit, the cost is verbosity but it's a lesser cost than production bug. (We already have some dicts in other place, but we want to have less of them over time).

Reply:
I agree that passing dicts around is bad in some cases, however I don't think having a model here would be benefitial, unless the type is exported as a model to be used by the consumer of the notifier data. Is that what you mean? You want these to be implemented with something like `pydantic` models?

Reply:
Ok, since it does not leave BackgroundHTTPNotifier we can live with the dict.

(I didn't mean pydantic, just dataclass, the type coercion of pydantic is not very helpful to avoid typing bugs :smile:)